system/api/create.go

```diff
@@ -242,3 +219,86 @@ func createContentHandler(res http.ResponseWriter, req *http.Request) {
 	}
 
 }
+
+func createContentHandlerJSON(res http.ResponseWriter, req *http.Request) (*url.Values, error) {
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	formData, err := jsonToURLValues(&body)
+	if err != nil {
+		return nil, err
+	}
+	return formData, nil
+}
+
+func createContentHandlerMultiPartForm(res http.ResponseWriter, req *http.Request) (*url.Values, error) {
+	err := req.ParseMultipartForm(1024 * 1024 * 4) // maxMemory 4MB
+	if err != nil {
+		return nil, err
+	}
+
+	ts := fmt.Sprintf("%d", int64(time.Nanosecond)*time.Now().UnixNano()/int64(time.Millisecond))
+	req.PostForm.Set("timestamp", ts)
+	req.PostForm.Set("updated", ts)
+
+	urlPaths, err := upload.StoreFiles(req)
+	if err != nil {
+		return nil, err
+	}
+
+	for name, urlPath := range urlPaths {
+		req.PostForm.Set(name, urlPath)
+	}
+
+	// check for any multi-value fields (ex. checkbox fields)
+	// and correctly format for db storage. Essentially, we need
+	// fieldX.0: value1, fieldX.1: value2 => fieldX: []string{value1, value2}
+	fieldOrderValue := make(map[string]map[string][]string)
+	for k, v := range req.PostForm {
+		if strings.Contains(k, ".") {
+			fo := strings.Split(k, ".")
+
+			// put the order and the field value into map
+			field := string(fo[0])
+			order := string(fo[1])
+			if len(fieldOrderValue[field]) == 0 {
+				fieldOrderValue[field] = make(map[string][]string)
+			}
+
+			// orderValue is 0:[?type=Thing&id=1]
+			orderValue := fieldOrderValue[field]
+			orderValue[order] = v
+			fieldOrderValue[field] = orderValue
+
+			// discard the post form value with name.N
+			req.PostForm.Del(k)
+		}
+
+	}
+
+	// add/set the key & value to the post form in order
+	for f, ov := range fieldOrderValue {
+		for i := 0; i < len(ov); i++ {
+			position := fmt.Sprintf("%d", i)
+			fieldValue := ov[position]
+
+			if req.PostForm.Get(f) == "" {
+				for i, fv := range fieldValue {
+					if i == 0 {
+						req.PostForm.Set(f, fv)
```

Comment:
Here we're mutating the request form, I didn't that in JSON handling. I seems a bad idea to change request for me. I would copy request data and work on it, which bring me doubts too.

I did this ` 	body, err := ioutil.ReadAll(req.Body) ` above, I'm afraid that subsequent call to `ioutil.ReadAll(req.Body)` may fail, I'm not really used to go readers yet... Will this prevent remaining code from getting the request body?